GUIDE.md

```diff
@@ -0,0 +1,299 @@
+if you want to use it in `App` struct.
+
+`App` creates a [`HandlerInput`](https://tg-rs.github.io/carapax/carapax/struct.HandlerInput.html) with `Context` and an `Update`,
+converts it into an input for a specific handler using `TryFromInput` trait.
+Handler will be executed only if `TryFromInput` returned `Some(T)`.
+So, in example above, `echo` will run only when `Update` contains any `Text` (message text, media captions).
+
+Use `Ref<T>` as argument type when you need to get an object from context.
+If object is not found, `TryFromInput` returns an error and handler will not run.
+
+### Chain
+
+Chain is a handler which allows to execute more than one handler:
+
+```rust
+use carapax::Chain;
+
+async fn handler1(_: ()) {}
+async fn handler2(_: ()) {}
+async fn handler3(_: ()) {}
+
+let chain = Chain::default()
+    .add_handler(handler1)
+    .add_handler(handler2)
+    .add_handler(handler3);
+```
+
+Handlers will run in same order as added.
+If a handler returns `HandlerResult::Stop` or `HandlerResult::Error(_)`, all handlers added after that handler will not run.
+
+### HandlerResult
+
+Normally every handler must return a `HandlerResult` or any type which converts into it:
+
+| From            | To                          |
+|-----------------|-----------------------------|
+| `()`            | `HandlerResult::Stop`       |
+| `true`          | `HandlerResult::Continue`   |
+| `false`         | `HandlerResult::Stop`       |
+| `Result<T, E>`  | `T.into::<HandlerResult>()` |
+
+
+## Predicates
+
+[`Predicate`](https://tg-rs.github.io/carapax/carapax/struct.Predicate.html)
+is a decorator which allows to decide, should handler run or not.
+It is useful when you need to implement a ratelimit or close a handler or entire bot from unwanted users and so on.
+
+A predicate handler must implement `Handler` trait, but return
+a [`PredicateResult`](https://tg-rs.github.io/carapax/carapax/enum.PredicateResult.html)
+or any type which converts into it:
+
+|   From            | To                            |
+|-------------------|-------------------------------|
+| `true`            | `PredicateResult::Continue`   |
+| `false`           | `PredicateResult::Stop`       |
+| `Result<T, E>`    | `T.into::<PredicateResult>()` |
+
+
+Example:
+
+```rust
+use carapax::{
+    methods::SendMessage,
+    types::{ChatId, Text},
+    Api, Chain, PredicateExt, Ref,
+};
+
+fn setup(chain: &mut Chain) {
+    let handler = pong.predicate(is_ping);
+    chain.add_handler(handler);
+}
+
+async fn is_ping(text: Text) -> bool {
+    text.data == "ping"
+}
+
+async fn pong(api: Ref<Api>, chat_id: ChatId) {
+    let method = SendMessage::new(chat_id, "pong");
+    api.execute(method).await.unwrap();
+}
+
+```
+
+### Commands
+
+[`CommandPredicate`](https://tg-rs.github.io/carapax/carapax/struct.CommandPredicate.html#)
+is a decorator which allows to run handlers only when user sent a command.
```

Comment:
```suggestion
is a decorator which allows to run a handler only if the update is a command.
```

Reply:
Update can not be a command, it contains a message with `bot_command` entity